src/Microsoft.AspNet.Cors/CorsMiddleware.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Cors.Core;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.WebUtilities;
+
+namespace Microsoft.AspNet.Cors
+{
+    /// <summary>
+    /// An ASP.NET middleware for handling CORS.
+    /// </summary>
+    public class CorsMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly CorsOptions _options;
+
+        /// <summary>
+        /// Instantiates a new <see cref="CorsMiddleware"/>.
+        /// </summary>
+        /// <param name="next"></param>
+        /// <param name="options"></param>
+        public CorsMiddleware(
+            [NotNull]RequestDelegate next,
+            [NotNull]CorsOptions options)
+        {
+            _next = next;
+            _options = options;
+
+            _options.PolicyProvider = _options.PolicyProvider ?? new CorsPolicyProvider();
+            _options.CorsEngine = _options.CorsEngine ?? new CorsEngine();
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            var corsRequestContext = new CorsRequestContext(context);
+
+            if (corsRequestContext.IsCorsRequest)
+            {
+                var corsPolicy = await _options.PolicyProvider.GetCorsPolicyAsync(corsRequestContext);
+
+                if (corsPolicy != null)
+                {
+                    var corsResult = _options.CorsEngine.EvaluatePolicy(corsRequestContext, corsPolicy);
+
+                    if (corsRequestContext.IsPreflight)
+                    {
+                        // Always respond to preflight requests.
+                        if (corsResult.IsValid)
+                        {
+                            context.Response.StatusCode = StatusCodes.Status200OK;
+                            WriteCorsHeaders(context, corsResult);
+                        }
+                        else
+                        {
+                            context.Response.StatusCode = StatusCodes.Status400BadRequest;
```

Comment:
When the preflight request is handled by MVC (in future) they will let the request flow to the next middleware in the pipeline if the action finder cannot find a matching action as requested by preflight request. But on middleware where there is just one policy it will handle it always.